From ecf71a880fa2838a8d302ebadce8d35a900ad86d Mon Sep 17 00:00:00 2001
From: Kees Cook <keescook@chromium.org>
Date: Wed, 22 May 2013 17:32:32 -0700
Subject: [PATCH] block: do not pass disk names as format strings
Git-commit: ecf71a880fa2838a8d302ebadce8d35a900ad86d
Patch-mainline: 3.11
References: bnc#822575 CVE-2013-2851

Disk names may contain arbitrary strings, so they must not be interpreted
as format strings. It seems that only md allows arbitrary strings to be
used for disk names, but this could allow for a local memory corruption
from uid 0 into ring 0.

CVE-2013-2851

Signed-off-by: Kees Cook <keescook@chromium.org>
Cc: stable@vger.kernel.org
Acked-by: Jan Kara <jack@suse.cz>

---
 block/genhd.c              |    2 +-
 drivers/block/nbd.c        |    3 ++-
 drivers/scsi/osd/osd_uld.c |    2 +-
 3 files changed, 4 insertions(+), 3 deletions(-)

Index: linux-3.7-openSUSE-12.3/block/genhd.c
===================================================================
--- linux-3.7-openSUSE-12.3.orig/block/genhd.c
+++ linux-3.7-openSUSE-12.3/block/genhd.c
@@ -517,7 +517,7 @@ static void register_disk(struct gendisk
 
 	ddev->parent = disk->driverfs_dev;
 
-	dev_set_name(ddev, disk->disk_name);
+	dev_set_name(ddev, "%s", disk->disk_name);
 
 	/* delay uevents, until we scanned partition table */
 	dev_set_uevent_suppress(ddev, 1);
Index: linux-3.7-openSUSE-12.3/drivers/block/nbd.c
===================================================================
--- linux-3.7-openSUSE-12.3.orig/drivers/block/nbd.c
+++ linux-3.7-openSUSE-12.3/drivers/block/nbd.c
@@ -685,7 +685,8 @@ static int __nbd_ioctl(struct block_devi
 			queue_flag_set_unlocked(QUEUE_FLAG_DISCARD,
 				nbd->disk->queue);
 
-		thread = kthread_create(nbd_thread, nbd, nbd->disk->disk_name);
+		thread = kthread_create(nbd_thread, nbd, "%s",
+					nbd->disk->disk_name);
 		if (IS_ERR(thread)) {
 			mutex_lock(&nbd->tx_lock);
 			return PTR_ERR(thread);
Index: linux-3.7-openSUSE-12.3/drivers/scsi/osd/osd_uld.c
===================================================================
--- linux-3.7-openSUSE-12.3.orig/drivers/scsi/osd/osd_uld.c
+++ linux-3.7-openSUSE-12.3/drivers/scsi/osd/osd_uld.c
@@ -465,7 +465,7 @@ static int osd_probe(struct device *dev)
 	oud->class_dev.class = &osd_uld_class;
 	oud->class_dev.parent = dev;
 	oud->class_dev.release = __remove;
-	error = dev_set_name(&oud->class_dev, disk->disk_name);
+	error = dev_set_name(&oud->class_dev, "%s", disk->disk_name);
 	if (error) {
 		OSD_ERR("dev_set_name failed => %d\n", error);
 		goto err_put_cdev;
